src/Kunstmaan/AdminBundle/EventListener/ToolbarListener.php

```diff
@@ -0,0 +1,131 @@
+    protected $authorizationChecker;
+
+    /** @var TokenStorage $tokenStorage */
+    protected $tokenStorage;
+
+    /** @var boolean $enabled */
+    protected $enabled;
+
+    /** @var boolean $debug */
+    protected $debug;
+
+    public function __construct(\Twig_Environment $twig, UrlGeneratorInterface $urlGenerator, DataCollector $dataCollector, AuthorizationChecker $authorizationChecker, TokenStorage $tokenStorage, $enabled, $debug)
+    {
+        $this->twig = $twig;
+        $this->urlGenerator = $urlGenerator;
+        $this->dataCollector = $dataCollector;
+        $this->authorizationChecker = $authorizationChecker;
+        $this->tokenStorage = $tokenStorage;
+        $this->enabled = $enabled;
+        $this->debug = $debug;
+    }
+
+    public function isEnabled()
+    {
+        return !$this->debug && $this->enabled;
+    }
+
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if ($this->isEnabled()) {
+            $response = $event->getResponse();
+            $request = $event->getRequest();
+            $url = $event->getRequest()->getRequestUri();
+            $token = $this->tokenStorage->getToken();
+
+            // do not capture redirects or modify XML HTTP Requests
+            if (!$event->isMasterRequest() || $request->isXmlHttpRequest() || $this->isAdminRoute($url) || !$token || !$this->authorizationChecker->isGranted('IS_AUTHENTICATED_FULLY')) {
+                return;
+            }
+
+            if ($response->isRedirection() || ($response->headers->has('Content-Type') && false === strpos($response->headers->get('Content-Type'), 'html'))
+                || 'html' !== $request->getRequestFormat()
+                || false !== stripos($response->headers->get('Content-Disposition'), 'attachment;')
+            ) {
+                return;
+            }
+
+            $this->injectToolbar($response, $request);
+        }
+    }
+
+    /**
+     * Injects the admin toolbar into the given Response.
+     *
+     * @param Response $response A Response instance
+     */
+    protected function injectToolbar(Response $response, Request $request)
+    {
+        $content = $response->getContent();
+        $pos = strripos($content, '</body>');
+
+        if (false !== $pos) {
+            $toolbar = "\n" . str_replace("\n", '', $this->twig->render(
+                    '@KunstmaanAdmin/Toolbar/toolbar.html.twig'
+                    , ['collectors' => $this->dataCollector->getDataCollectors()])) . "\n";
+            $content = substr($content, 0, $pos) . $toolbar . substr($content, $pos);
+            $response->setContent($content);
+        }
+    }
+
+    /**
+     * @param string $url
+     *
+     * @return bool
+     */
+    private function isAdminRoute($url)
+    {
+        //If the url contains an admin part and a preview part then it is not an admin route
+        preg_match('/^(\/app_[a-zA-Z]+\.php)?\/([a-zA-Z_-]{2,5}\/)?admin(\/.*)?\/preview/', $url, $matches);
```

Comment:
The bundles code contains more code like this on checking of the admin route. You can always make a pull request with a more abstract way of doing this.